### PR TITLE
Provide feedback if it takes a while to download a remote component on `spin up`

### LINKED ISSUE #2109: Improve experience if it takes a while to download a "remote" component on `spin up`
I was on an airplane, and I decided to run a Spin app using the fileserver, with a reference to Github in spin.toml. Now that took a while, as the bits of the fileserver was chasing the airplane across the Atlantic (I'm not sure they ever caught up). In those weird cases, it would be great if at `spin up` a note is fed back to the use, that Spin is currently busy downloading that remote component, so that it's easy to understand what we'er waiting for.

### PR DESCRIPTION
**Pull request by Archisman-Mridha**

### CODE REVIEW

**Review by itowlson — CHANGES_REQUESTED** [review]
Hi @Archisman-Mridha, welcome and thanks!

I'm afraid this PR doesn't answer the issue you mentioned (unless I've misread the link), and it's not quite the way we usually do these "taking a long time" warnings anyway.  What you have here would _always_ print a message whenever an application was pulled or run from an OCI registry reference.  What #2109 is looking for is to print a message for a **remote component**, and to print it **if it takes a while**.  Let me drill into those two phrases a bit:

* By a "remote component", we mean a _component_ referenced via a remote reference.  Today, the only remote references for components are HTTP references such as the one in the `static-fileserver` template.  The emphasis on _component_ here is because such a component could be in a _local_ application, even a freshly created one.  (It's more surprising when a local app is slow to load than when a remote app is slow to load!)  This is a different code path from pulling an entire from the registry to local - see e.g. https://github.com/fermyon/spin/blob/main/crates/loader/src/http.rs.

* By "if it takes a while", we mean we don't print "downloading / starting / whatever" _every_ time the user does the thing.  This risks the user tuning out the warning that they see all the time, so not registering it when it actually matters.  Another example of this is if an app's Wasm modules are taking a long time to load.  In this case we wait 1.25 seconds before displaying a "slow load" message (https://github.com/fermyon/spin/blob/7a6b81b2545029ec90d49a6e00ca20dc31843535/crates/trigger/src/cli.rs#L191).  The `spin_common::sloth` module provides a function which manages this by running a background timer and printing only if the returned guard is still live when the timer expires.

I hope this is useful feedback - of course please let me know if the desired behaviour or the way to go about it is not clear.  And thanks again for getting involved!

**Comment by itowlson** [comment]
Oops.  Thanks for the catch on the title @Archisman-Mridha...!  _embarrassed_

**Comment by Archisman-Mridha** [comment]
@itowlson Thanks for the quick and in-detail response! Sorry for misunderstanding the issue. I am working on it....

## Steps I took to clear out my concepts

- I read about the [WASM Component Model](https://www.fermyon.com/blog/webassembly-component-model) and understood what are `WASM components`.

- Then I read [what is and why Spin](https://www.fermyon.com/blog/introducing-spin) ?

- Then I read about the [architecture of a Spin application](https://developer.fermyon.com/spin/v2/writing-apps) and understood what we mean by `components in a Spin application`.

- I investigated the template at `templates/static-fileserver/content` and also the [Spin application manifest reference](https://developer.fermyon.com/spin/v2/manifest-reference).

## My current understanding

> A spin application contains multiple WASM components. These WASM components can also be located in a remote fileserver. When we execute `spin up`, these WASM components are download to the local machine.

 Components downloading is done by the `verified_download` function located at - `crates/loader/src/http.rs`. I just need to put a `sloth guard` inside `verified_download`.

Am I right ?

**Comment by itowlson** [comment]
Your understanding is right.  Your proposal is almost right, but there is a small nuance: we might now want to place the sloth guard _inside_ `verified_download`, because then an app with more than one remote component would print the warning several times.  This might be okay with clear messaging about which components are taking time; but more usually we place the guard around the entire process (of loading, downloading, etc.), so that it is printed only once.  (Again, see the sloth guard around loading Wasm modules, where we print a single "loading is taking a while" rather than listing each component.)

Why do it this way?  Consider a manifest with a large number of remote components.  (For example, `redirect` is a remote component.  A site might have a lot of redirects.)  Printing a message for each of them would produce a lot of console spew for the user to burrow through.  Worse, if each component takes, say, 1.1 seconds to load, but in total they take 11 seconds to load, we don't see the feedback at all!

So a more desirable plan is to find where _all_ the remote components are being downloaded (e.g. in a loop or as a map of futures), which will be a place that directly or indirectly calls `verified_download`.  Then the sloth guard can be placed around that loop or those futures or that function or whatever.

Does that help?  Again, happy to discuss further or give you more specific pointers if that would be useful.

**Comment by Archisman-Mridha** [comment]
@itowlson Thanks for being supportive!

I figured it this is where all the components are loaded concurrently : `crates/loader/src/local.rs`, **line number 97 - 103** :

```rust
// Load all components concurrently
let components = try_join_all(components.into_iter().map(|(id, c)| async move {
    self.load_component(&id, c)
        .await
        .with_context(|| format!("Failed to load component `{id}`"))
}))
.await?;
```

I need to put the `sloth guard` somewhere here. Am I right ?

**Comment by Archisman-Mridha** [comment]
Good morning :). I pushed a commit. Please let me know if the changes look okay or not.

**Review by itowlson — CHANGES_REQUESTED** [review]
The approach looks good - thanks for persevering with this!  However, there is an important detail about keeping the `SlothGuard` alive (see comments).  (This is on us - Rust has a way to warn about this kind of thing, and I forgot to set it up.)  Other than that and a couple of nits it looks good to go though!

File: crates/loader/src/local.rs [hunk 1]
```diff
@@ -94,6 +94,8 @@ impl LocalLoader {
             })
             .collect::<Result<Vec<_>>>()?;
 
+        warn_if_component_load_slothful();
```
[the remark this answers is not in the document]
Comment (Archisman-Mridha) [reply crates/loader/src/local.rs:97]: How did I not catch that! This makes me worried.

**Review by itowlson — APPROVED** [review]
This looks good to me!  I will merge once CI completes (barring any hiccups).  Thanks for your patience and perseverance.

**Comment by itowlson** [comment]
Oh!  Could you mark the PR ready for review please @Archisman-Mridha?  It's still draft and can't be merged in that state.

**Comment by Archisman-Mridha** [comment]
> Oh! Could you mark the PR ready for review please @Archisman-Mridha? It's still draft and can't be merged in that state.

Done

**Pull request merged**